presto-main/src/main/java/io/prestosql/server/DynamicFilterService.java

```diff
@@ -0,0 +1,268 @@
+    {
+        collectDynamicFiltersExecutor.scheduleWithFixedDelay(this::collectDynamicFilters, 0, taskConfig.getStatusRefreshMaxWait().toMillis(), TimeUnit.MILLISECONDS);
+    }
+
+    public void registerQuery(QueryId queryId, Supplier<List<StageInfo>> stageInfoSupplier)
+    {
+        queries.putIfAbsent(queryId, stageInfoSupplier);
+    }
+
+    @VisibleForTesting
+    public void collectDynamicFilters()
+    {
+        List<QueryId> queriesWithDynamicFilters = dynamicFilterSummaries.entrySet()
+                .stream()
+                .filter(entry -> !entry.getValue().isDone())
+                .map(entry -> entry.getKey().getQueryId())
+                .distinct()
+                .collect(toImmutableList());
+
+        for (QueryId queryId : queriesWithDynamicFilters) {
+            Supplier<List<StageInfo>> stageInfoSupplier = queries.get(queryId);
+            if (stageInfoSupplier != null) {
+                for (StageInfo stageInfo : getStagesWithDynamicFilters(stageInfoSupplier.get())) {
+                    StageState stageState = stageInfo.getState();
+                    // wait until stage has finished scheduling tasks
+                    if ((stageState == SCHEDULING_SPLITS || stageState == SCHEDULED || stageState == RUNNING || stageState == FINISHED)) {
+                        List<TaskInfo> tasks = stageInfo.getTasks();
+                        Map<String, ImmutableList<TupleDomain<String>>> stageDynamicTupleDomains = tasks.stream()
+                                .map(TaskInfo::getDynamicTupleDomains)
+                                .flatMap(taskDynamicTupleDomains -> taskDynamicTupleDomains.entrySet().stream())
+                                .collect(groupingBy(Map.Entry::getKey, mapping(Map.Entry::getValue, toImmutableList())));
+                        for (Map.Entry<String, ImmutableList<TupleDomain<String>>> entry : stageDynamicTupleDomains.entrySet()) {
+                            List<TupleDomain<String>> sourceDynamicTupleDomains = entry.getValue();
+                            // check if all tasks of a dynamic filter source have reported TupleDomain summary
+                            if (tasks.size() == sourceDynamicTupleDomains.size()) {
+                                dynamicFilterSummaries.computeIfAbsent(
+                                        SourceDescriptor.of(queryId, entry.getKey()),
+                                        ignored -> SettableFuture.create())
+                                        .set(columnWiseUnion(sourceDynamicTupleDomains));
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    public ListenableFuture<TupleDomain<String>> getSummary(QueryId queryId, String source)
+    {
+        return dynamicFilterSummaries.computeIfAbsent(SourceDescriptor.of(queryId, source), ignored -> SettableFuture.create());
+    }
+
+    public void removeQuery(QueryId queryId)
```

Comment:
listeners are called asynchronously, but you update two collections without locks. I think you should just `synchronize` here and not use concurrent collections

Reply:
Switched to using synchronized methods without concurrent collections in `DynamicFilterService`